Anchor Gutenboarding: Query the WPCOM backend for matching anchor sites and redir

mreishus opened the pull request:
#### Changes proposed in this Pull Request

* The "Gather Intent" step of anchor flavored gutenboarding now checks to see if you already have an existing anchor site matching the params, and if so, skips gutenboarding and redirects your browser there.

![2021-02-01_17-20](https://user-images.githubusercontent.com/937354/106530431-e9d56f00-64b1-11eb-826d-f360e71fa42d.png)


#### Testing instructions

* Apply D56339-code
* Start Calypso
* Part 1: New user
  * Open an incognito window and visit a "magic gutenboarding podcast url"
  * Register a new account and go through the gutenboarding process. Note the account email and password.
  * You should end up on a newly created site with anchor content in it and be the exact same as before.
* Part 2: Returning User (Already Logged In)
  * If you still have the incognito window used before that's logged in, make a new tab.  If you don't, visit http://calypso.localhost:3000 and log in to this account.
  * In your already logged in state, visit Magic Gutenboarding Podcast Urls. 
    * If you visit one for the same podcast, you should start to go through the gutenboarding process, but instead of making a new site, you're redirected to your already existing site.
    * If you visit one for a podcast you haven't made yet, you should go through the gutenboarding process again and make a second site.
* Part 3: Returning User (Not logged in, log in during Gutenboarding)
  * Close all incognito windows and open a new one, to guarantee you'll be logged out.
  * In this window, visit a magic gutenboarding podcast url.  
  * You should see the register screen.  Here, log in instead.
![2021-02-01_17-15](https://user-images.githubusercontent.com/937354/106530050-33718a00-64b1-11eb-9e2c-821a80452f16.png)
  * You should skip the gutenboarding process and be taken to your site. (If you are using gutenboarding url corresponding to an anchor podcast that you've already made a site for.)

* Magic Gutenboarding Podcast Urls
    * "Brave Not Perfect": http://calypso.localhost:3000/new?anchor_podcast=22b6608&anchor_episode=e324a06c-3148-43a4-85d8-34c0d8222138&spotify_url=https%3A%2F%2Fopen.spotify.com%2Fshow%2F6HTZdaDHjqXKDE4acYffoD%3Fsi%3DEVfDYETjQCu7pasVG5D73Q&flags=anchor-fm-dev

Related to 439-gh-Automattic/dotcom-manage

mreishus committed:
Anchor Gutenboarding: Query the WPCOM backend for matching anchor sites and redir

kwight commented on client/landing/gutenboarding/hooks/use-detect-matching-anchor-site.ts:
```diff
@@ -0,0 +1,55 @@
+		const anchorEndpointUrl = addQueryArgs( anchorEndpointBase, queryParts );
+
+		// Hit Endpoint
+		wpcom.req
```
In general, we try to keep these calls all in the same place here: client/lib/wpcom-undocumented/lib/undocumented.js

mreishus replied:
Extracted here https://github.com/Automattic/wp-calypso/pull/49571/commits/eb52fc65ab59d19ea5a434d54f65bec74abb3b0a

kwight commented:
I wasn't able to get past the first step, receiving an auth error on site creation:

<img width="909" alt="Screen Shot 2021-02-02 at 7 56 29 AM" src="https://user-images.githubusercontent.com/349751/106626162-42e1e900-652c-11eb-979c-ce261e44706c.png">

mreishus committed:
Extract getMatchingAnchorSite() to undocumented.js

mreishus commented:
> I wasn't able to get past the first step, receiving an auth error on site creation:

#### Browser Issue
I'm able to replicate the XHR to `/rest/v1.1/sites/new` failing in Chrome, but not in Firefox (in Firefox, I can successfully make the site).  
#### Not unique to this branch
I'm also seeing the same error happen on the `trunk` branch of calypso with a clean sandbox.
#### Unknown Cause
I don't know the cause of it.  I have seen some similar odd behavior happening in calypso dev environments inside chrome only. See  https://github.com/Automattic/wp-calypso/issues/46127 and p1606322020361400-slack-CJS75TX3R . Perhaps some sort of wider issue between recent chrome changes, calypso and the wpcom api?

mreishus committed:
Solve flash of aquire intent page for existing anchor users

- New anchor users will see a small delay before the aquire intent page
  renders, while we check the backend for existing sites.

kwight commented:
This works great! 🎉 

I used a regular window to avoid any possible incognito+cookies issues, but had no problem with the default Gutenboarding flow, the Anchor new user and returning flows, plus giving a specific episode ID – lovely.

mreishus committed:
Use addQueryArgs from calypso/lib, not @wordpress/url

Old way was producing error:

Time=2021-02-03T21:31:59.670Z | $ node bin/copy-production-modules.js
Time=2021-02-03T21:31:59.809Z | Error: package not found: react-native, dependency of ./node_modules/react-native-url-polyfill
    at processPackage (/calypso/bin/copy-production-modules.js:99:10)
    at processPackage (/calypso/bin/copy-production-modules.js:109:3)
    at processPackages (/calypso/bin/copy-production-modules.js:38:3)
    at Object.<anonymous> (/calypso/bin/copy-production-modules.js:21:26)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47

Sergio:

@wordpress/url, which depends on react-native-url-polyfill, and that has
a peerDependency and a devDependency on react-native

aha! when building the server bundle we copy all the required modules,
including peerDependencies. react-native-url-polyfill declares a peer
dependency on react-native, but we have not fullfilled it (i.e. we don’t
have react-native in the repo)

mreishus commented:
* Changed `import { addQueryArgs } from '@wordpress/url';` to `import addQueryArgs from 'calypso/lib/url/add-query-args';` and re-ordered arguments
  * `@wordpress/url` was bringing in react-native and breaking the build.
* New version of addQueryArgs is causing TS errors.
  * Fixing typescript errors: Waiting for [Change import to import type in client/lib/url helpers] #49687

Waiting for [Change import to import type in client/lib/url helpers] #49687 to proceed

mreishus committed:
Remove addQueryArgs, not needed

mreishus commented:
e2e tests are a known failure: p1612457621019500/1612456065.018600-slack-C1A1EKDGQ
